src/input/tablet.c

```diff
@@ -83,30 +92,76 @@ adjust_for_rotation(enum rotation rotation, double *x, double *y)
 	}
 }
 
+static void
+adjust_for_rotation_relative(enum rotation rotation, double *dx, double *dy)
+{
+	double tmp;
+	switch (rotation) {
+	case LAB_ROTATE_NONE:
+		break;
+	case LAB_ROTATE_90:
+		tmp = *dx;
+		*dx = -*dy;
+		*dy = tmp;
+		break;
+	case LAB_ROTATE_180:
+		*dx = -*dx;
+		*dy = -*dy;
+		break;
+	case LAB_ROTATE_270:
+		tmp = *dx;
+		*dx = *dy;
+		*dy = -tmp;
+		break;
+	}
+}
+
+static void
+adjust_for_motion_sensitivity(double motion_sensitivity, double *dx, double *dy)
+{
+	*dx = *dx * motion_sensitivity;
+	*dy = *dy * motion_sensitivity;
+}
+
 static struct wlr_surface*
-tablet_get_coords(struct drawing_tablet *tablet, double *x, double *y)
+tablet_get_coords(struct drawing_tablet *tablet, double *x, double *y, double *dx, double *dy)
 {
 	*x = tablet->x;
 	*y = tablet->y;
+	*dx = tablet->dx;
+	*dy = tablet->dy;
 	adjust_for_tablet_area(tablet->tablet->width_mm, tablet->tablet->height_mm,
 		rc.tablet.box, x, y);
 	adjust_for_rotation(rc.tablet.rotation, x, y);
+	adjust_for_rotation_relative(rc.tablet.rotation, dx, dy);
+	adjust_for_motion_sensitivity(rc.tablet_tool.relative_motion_sensitivity, dx, dy);
 
 	if (rc.tablet.force_mouse_emulation
 			|| !tablet->tablet_v2) {
 		return NULL;
 	}
 
-	/* Convert coordinates: first [0, 1] => layout, then layout => surface */
-	double lx, ly;
-	wlr_cursor_absolute_to_layout_coords(tablet->seat->cursor,
-		tablet->wlr_input_device, *x, *y, &lx, &ly);
+	/* convert coordinates: first [0, 1] => layout, then layout => surface */
+
+	/* initialize here to avoid a maybe-uninitialized warning */
+	double lx = -1, ly = -1;
+	switch (tablet->motion_mode) {
+	case LAB_TABLET_MOTION_ABSOLUTE:
+		wlr_cursor_absolute_to_layout_coords(tablet->seat->cursor,
+			tablet->wlr_input_device, *x, *y, &lx, &ly);
+		break;
+	case LAB_TABLET_MOTION_RELATIVE:
+		/* dx,dy are already layout coordinate deltas */
```

Comment:
Note to self, not really true https://wayland.freedesktop.org/libinput/doc/latest/api/group__event__tablet.html#ga868763e264a48cb8df72f5b4724a3c4e